src/Swarm/Game/Scenario/Objective/Graph.hs

```diff
@@ -0,0 +1,140 @@
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- | Utilities for performing graph analysis on Objective prerequisites
+module Swarm.Game.Scenario.Objective.Graph where
+
+import Control.Arrow ((&&&))
+import Data.Aeson
+import Data.BoolExpr (Signed (Positive))
+import Data.BoolExpr qualified as BE
+import Data.Graph (Graph, SCC (AcyclicSCC), graphFromEdges, stronglyConnComp)
+import Data.Map (Map)
+import Data.Map.Strict qualified as M
+import Data.Maybe (mapMaybe)
+import Data.Set qualified as Set
+import Data.Text qualified as T
+import Data.Tuple (swap)
+import GHC.Generics (Generic)
+import Swarm.Game.Scenario.Objective
+import Swarm.Game.Scenario.Objective.Logic as L
+import Swarm.Game.Scenario.Objective.WinCheck
+
+-- | This is only needed for constructing a Graph,
+-- which requires all nodes to have a key.
+data ObjectiveId
+  = Label (Signed ObjectiveLabel)
+  | -- | for unlabeled objectives
+    Ordinal Int
+  deriving (Eq, Ord, Show, Generic, ToJSON)
+
+data GraphInfo = GraphInfo
+  { actualGraph :: Graph
+  , isAcyclic :: Bool
+  , sccInfo :: [SCC Objective]
+  , nodeIDs :: [ObjectiveId]
+  }
+  deriving (Show, Generic, ToJSON)
+
+instance ToJSON (SCC Objective) where
+  toJSON = String . T.pack . show
+
+instance ToJSON Graph where
+  toJSON = String . T.pack . show
+
+getConstFromSigned :: BE.Signed a -> a
```

Comment:
Don't want to add to `Data.BoolExpr` at this point.  Maybe if we take over its upstream maintenance.

Reply:
Indeed, I guess I meant more something like `Data.BoolExpr.Simplify`.  (See also my comment on renaming the `Swarm.Game.Scenario.Objective.Simplify` module to something like `Data.BoolExpr.Simplify`, since none of the code there is specific to scenarios or objectives at all.)